plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.lang.Math.max;
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * Abstract base for scrolling across a search and executing bulk actions on all
+ * results.
+ */
+public abstract class AbstractAsyncBulkByScrollAction<Request extends AbstractBulkByScrollRequest<Request>, Response> {
+    protected final Request mainRequest;
+
+    private final AtomicLong total = new AtomicLong(-1);
+    private final AtomicLong startTime = new AtomicLong(-1);
+    private final AtomicLong updated = new AtomicLong(0);
+    private final AtomicLong created = new AtomicLong(0);
+    private final AtomicLong deleted = new AtomicLong(0);
+    private final AtomicInteger batches = new AtomicInteger(0);
+    private final AtomicLong versionConflicts = new AtomicLong(0);
+    private final AtomicReference<String> scroll = new AtomicReference<>();
+    private final List<Failure> failures = new CopyOnWriteArrayList<>();
+    private final Set<String> destinationIndices = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+
+    private final ESLogger logger;
+    private final Client client;
+    private final ThreadPool threadPool;
+    private final SearchRequest firstSearchRequest;
+    private final ActionListener<Response> listener;
+
+    public AbstractAsyncBulkByScrollAction(ESLogger logger, Client client, ThreadPool threadPool, Request mainRequest,
+            SearchRequest firstSearchRequest, ActionListener<Response> listener) {
+        this.logger = logger;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.mainRequest = mainRequest;
+        this.firstSearchRequest = firstSearchRequest;
+        this.listener = listener;
+    }
+
+    protected abstract BulkRequest buildBulk(Iterable<SearchHit> docs);
+
+    protected abstract Response buildResponse(long took);
+
+    public void start() {
+        initialSearch();
+    }
+
+    /**
+     * Count of documents updated.
+     */
+    public long updated() {
+        return updated.get();
+    }
+
+    /**
+     * Count of documents created.
+     */
+    public long created() {
+        return created.get();
+    }
+
+    /**
+     * Count of successful delete operations.
+     */
+    public long deleted() {
+        return deleted.get();
+    }
+
+    /**
+     * The number of scan responses this request has processed.
+     */
+    public int batches() {
+        return batches.get();
+    }
+
+    public long versionConflicts() {
+        return versionConflicts.get();
+    }
+
+    public long successfullyProcessed() {
+        return updated.get() + created.get() + deleted.get();
+    }
+
+    public List<Failure> failures() {
+        return unmodifiableList(failures);
+    }
+
+    void initialSearch() {
```

Comment:
Its pretty common for the handles to have package visibility to prevent the creation of those accessors for the inner class listeners. But in this case it isn't used so I'll set it to private.